### PR TITLE
[ONNX] Slice operator

### PR DESCRIPTION
**Pull request by Deliadd**

### CODE REVIEW

**Comment by Deliadd** [comment]
write test_loading_model similar to pytorch_reshape, still have some questions unsolved.
slice has the similar function as torch.tensor.narrow (according to document)

**Comment by Deliadd** [comment]
model = Slice()
args = tensor([[[-1.3029, -0.6377],
         [ 1.5004, -1.2538]],

        [[ 1.2674,  1.2191],
         [ 0.7963, -0.4275]]])
f = '/tmp/tmpjyupfxif/085428be0d9946a68d51bff3c3929fa1.onnx'
export_params = True, verbose = False, training = False, input_names = None
output_names = None, operator_export_type = OperatorExportTypes.ONNX
export_type = 1, example_outputs = None, propagate = False

    def _export(model, args, f, export_params=True, verbose=False, training=False,
                input_names=None, output_names=None, operator_export_type=OperatorExportTypes.ONNX,
                export_type=ExportTypes.PROTOBUF_FILE, example_outputs=None, propagate=False):
        graph, params, torch_out = _model_to_graph(model, args, f, verbose,
                                                   training, input_names,
                                                   output_names, operator_export_type,
                                                   example_outputs, propagate)
    
        # TODO: Don't allocate a in-memory string for the protobuf
        from torch.onnx.symbolic import _onnx_opset_version
        defer_weight_export = export_type is not ExportTypes.PROTOBUF_FILE
        if export_params:
>           proto, export_map = graph.export(params, _onnx_opset_version, defer_weight_export, operator_export_type)
E           RuntimeError: ONNX export failed: Couldn't export operator aten::narrow

**Comment by Deliadd** [comment]
model = Slice()
args = tensor([[[-1.3029, -0.6377],
         [ 1.5004, -1.2538]],

        [[ 1.2674,  1.2191],
         [ 0.7963, -0.4275]]])
f = '/tmp/tmpjyupfxif/085428be0d9946a68d51bff3c3929fa1.onnx'
export_params = True, verbose = False, training = False, input_names = None
output_names = None, operator_export_type = OperatorExportTypes.ONNX
export_type = 1, example_outputs = None, propagate = False

    def _export(model, args, f, export_params=True, verbose=False, training=False,
                input_names=None, output_names=None, operator_export_type=OperatorExportTypes.ONNX,
                export_type=ExportTypes.PROTOBUF_FILE, example_outputs=None, propagate=False):
        graph, params, torch_out = _model_to_graph(model, args, f, verbose,
                                                   training, input_names,
                                                   output_names, operator_export_type,
                                                   example_outputs, propagate)
    
        # TODO: Don't allocate a in-memory string for the protobuf
        from torch.onnx.symbolic import _onnx_opset_version
        defer_weight_export = export_type is not ExportTypes.PROTOBUF_FILE
        if export_params:
>           proto, export_map = graph.export(params, _onnx_opset_version, defer_weight_export, operator_export_type)
E           RuntimeError: ONNX export failed: Couldn't export operator aten::narrow
E           
E           Defined at:
E           /home/dingdi/analytics-zoo/pyzoo/test/zoo/pipeline/onnx/test_model_loading.py(880): forward
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/nn/modules/module.py(465): _slow_forward
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/nn/modules/module.py(475): __call__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/jit/__init__.py(109): forward
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/nn/modules/module.py(477): __call__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/jit/__init__.py(77): get_trace_graph
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/onnx/utils.py(144): _trace_and_get_graph_from_model
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/onnx/utils.py(177): _model_to_graph
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/onnx/utils.py(226): _export
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/onnx/utils.py(94): export
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/torch/onnx/__init__.py(26): export
E           /home/dingdi/analytics-zoo/pyzoo/test/zoo/pipeline/utils/test_utils_onnx.py(37): dump_pytorch_to_onnx
E           /home/dingdi/analytics-zoo/pyzoo/test/zoo/pipeline/utils/test_utils_onnx.py(58): compare_with_pytorch
E           /home/dingdi/analytics-zoo/pyzoo/test/zoo/pipeline/onnx/test_model_loading.py(884): test_onnx_slice
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/unittest/case.py(605): run
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/unittest/case.py(653): __call__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/unittest.py(176): runtest
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/runner.py(112): pytest_runtest_call
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(614): execute
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(265): __init__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(248): _wrapped_call
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(613): execute
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(334): <lambda>
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(339): _hookexec
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(745): __call__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/runner.py(182): <lambda>
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/runner.py(196): __init__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/runner.py(182): call_runtest_hook
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/runner.py(162): call_and_report
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/runner.py(82): runtestprotocol
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/runner.py(68): pytest_runtest_protocol
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(614): execute
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(265): __init__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(248): _wrapped_call
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(613): execute
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(265): __init__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(248): _wrapped_call
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(613): execute
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(334): <lambda>
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(339): _hookexec
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(745): __call__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/main.py(169): pytest_runtestloop
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(614): execute
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(334): <lambda>
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(339): _hookexec
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(745): __call__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/main.py(146): _main
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/main.py(110): wrap_session
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/main.py(139): pytest_cmdline_main
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(614): execute
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(334): <lambda>
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(339): _hookexec
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py(745): __call__
E           /home/dingdi/anaconda3/envs/py35/lib/python3.5/site-packages/_pytest/config.py(58): main
E           /home/dingdi/pycharm-community-2017.2.2/helpers/pycharm/_jb_pytest_runner.py(31): <module>
E           
E           
E           Graph we tried to export:
E           graph(%0 : Float(2, 2, 2)) {
E             %1 : Float(2, 2, 2) = aten::narrow[dim=0, start=0, length=2](%0), scope: Slice
E             return (%1);
E           }

File: pyzoo/zoo/pipeline/api/onnx/mapper/slice.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    def _to_tensor(self):
+        lens = []
+        input = self.model_inputs[0].zvalue
+        ends = self.onnx_attr['ends']
+        starts = self.onnx_attr['starts']
+        if "axes" in self.onnx_attr.keys():
+            axes = self.onnx_attr['axes']
+        else:
+            axes = []
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/slice.py]: how about `axes = range(len(starts))`?

File: pyzoo/zoo/pipeline/api/onnx/mapper/slice.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            for q in range(len(starts)):
+                axes.append(q)
+        for j in range(len(starts)):
+            lens.append(ends[j] - starts[j])
+            if lens[j] < 0:
+                lens[j] -= 1
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/slice.py]: ?

Comment (Deliadd) [reply pyzoo/zoo/pipeline/api/onnx/mapper/slice.py]: different definitions about the boundary, so I adjust the include value.

File: pyzoo/zoo/pipeline/api/onnx/mapper/slice.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        if "axes" in self.onnx_attr.keys():
+            axes = self.onnx_attr['axes']
+        else:
+            axes = []
+            for q in range(len(starts)):
+                axes.append(q)
+        for j in range(len(starts)):
+            lens.append(ends[j] - starts[j])
+            # slice(, len=-1) equals to slice(, len=length)
+            # y = x[0:2, 0:-2] means start is（0，0）, ends is（2，-2）, lens is(2, -3)
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/slice.py]: y = x[0:2, 0:-1] means start is（0，0）, ends is（2，-1）, which is equivalent to slice(, len=-2) as "end=-1" is exclusive here.

**Comment by dingdi21151626** [comment]
Add some comments. Jenkins passed. ID is 1419.

**Comment by Deliadd** [comment]
Add some comments. Jenkins passed. ID is 1419.

**Pull request merged**